File: app/describer.py
import os
import json
import asyncio
import threading
import requests
import time

def _post_with_retry(url, headers, json_payload, max_retries=4, backoff=3):
    """
    POST with retry + automatic Groq key rotation on 429.
    `headers` must contain the Authorization key; it will be mutated on rotation.
    """
    for attempt in range(max_retries):
        try:
            response = requests.post(url, headers=headers, json=json_payload, timeout=10)
            if response.status_code == 429:
                next_key = groq_pool.rotate()
                if next_key:
                    headers["Authorization"] = f"Bearer {next_key}"
                if attempt < max_retries - 1:
                    time.sleep(backoff * (attempt + 1))
                    continue
            elif response.status_code in [500, 502, 503, 504]:
                if attempt < max_retries - 1:
                    time.sleep(backoff * (attempt + 1))
                    continue
            return response
        except (requests.exceptions.ConnectionError, requests.exceptions.Timeout) as e:
            if attempt < max_retries - 1:
                time.sleep(backoff)
            else:
                raise e

def load_env():
    """Custom light-weight dotenv loader for zero external dependencies."""
    env_path = os.path.join(os.path.dirname(__file__), ".env")
    if os.path.exists(env_path):
        with open(env_path, "r", encoding="utf-8") as f:
            for line in f:
                line = line.strip()
                if "=" in line and not line.startswith("#"):
                    k, v = line.split("=", 1)
                    os.environ[k.strip()] = v.strip()

# Load env immediately
load_env()

# ── Groq API Key Pool ──────────────────────────────────────────────────────
class _GroqKeyPool:
    """
    Round-robin key pool for Groq API keys.
    On a 429 rate-limit response the caller advances to the next key automatically.
    
    .env configuration (add as many keys as you have):
        GROQ_API_KEY=primary_key
        GROQ_API_KEYS=primary_key,second_key,third_key   # optional pool
    If only GROQ_API_KEY is set it works as a single-key pool (original behaviour).
    """
    def __init__(self):
        self._lock = threading.Lock()
        self._index = 0
        self._keys: list[str] = []

    def _load(self) -> list[str]:
        """Read keys fresh from env every time (picks up server restarts / reloads)."""
        pool_str = os.environ.get("GROQ_API_KEYS", "")
        if pool_str:
            keys = [k.strip() for k in pool_str.split(",") if k.strip()]
        else:
            keys = []
        primary = os.environ.get("GROQ_API_KEY", "").strip()
        if primary and primary not in keys:
            keys.insert(0, primary)
        return keys

    def current(self) -> str:
        """Return the currently active key."""
        self._keys = self._load()
        if not self._keys:
            return ""
        with self._lock:
            return self._keys[self._index % len(self._keys)]

    def rotate(self) -> str:
        """Advance to the next key (called on 429) and return it."""
        self._keys = self._load()
        if not self._keys:
            return ""
        with self._lock:
            self._index = (self._index + 1) % len(self._keys)
            return self._keys[self._index]

    def __len__(self):
        return len(self._load())


groq_pool = _GroqKeyPool()

def generate_column_insights(col_name, profile_details, dataset_context=""):
    """
    Acts like an expert Senior Data Analyst. Calls the lightning-fast Groq API 
    (Llama3-8b) to generate JSON-formatted business descriptions and cleaning recommendations.
    """
    api_key = groq_pool.current()
    if not api_key:
        return {
            "description": "API Key missing. Please set GROQ_API_KEY in the app/.env file.",
            "recommendation": "Set your Groq API key to unlock expert AI descriptions."
        }

    url = "https://api.groq.com/openai/v1/chat/completions"
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    # Prepare detailed statistical profile text
    stats_text = (
        f"Column Name: '{col_name}'\n"
        f"Basic Pandas Type: {profile_details.get('pandas_dtype')}\n"
        f"Inferred Semantic Type: {profile_details.get('semantic_type')}\n"
        f"Completeness: {profile_details.get('non_null_count')} valid records, "
        f"{profile_details.get('null_percentage')}% missing.\n"
        f"Cardinality: {profile_details.get('unique_count')} unique values ({profile_details.get('unique_ratio')}% ratio).\n"
        f"Sample values: [{profile_details.get('sample_data')}]\n"
    )
    
    if profile_details.get("mean") is not None:
        stats_text += (
            f"Numeric Distribution: Min={profile_details.get('min')}, Max={profile_details.get('max')}, "
            f"Mean={profile_details.get('mean')}, StdDev={profile_details.get('std')}.\n"
            f"Anomalies: {profile_details.get('outliers_count')} outliers detected via Interquartile Range (IQR).\n"
        )

    system_prompt = (
        "You are a world-class Principal Data Architect and Senior Data Analyst with 15+ years of experience in enterprise business intelligence, database indexing, and data governance.\n"
        "Your objective is to analyze a column's statistical profile, cardinality, null ratios, anomalies, and sample values, and write authoritative, elite-level business summaries.\n"
        "You must output your response in EXACT JSON format with these two fields:\n"
        "1. 'description': A highly professional, single-sentence executive business definition. Explain *why* this data matters in a business context, what business KPIs it drives, and what it represents. Use sophisticated, concise terminology (e.g., 'chronological anchor', 'categorical dimension', 'metric variable') instead of basic descriptions.\n"
        "2. 'recommendation': An expert data quality, cleaning, and optimization recommendation. Provide concrete, advanced technical instructions based on the column's metrics (e.g., specify how to handle the exact null percentage, whether to index it for faster lookups, how to treat the outliers, or suggesting binning, scaling, or database partitioning based on cardinality and type).\n"
        "Your output must be valid JSON only, without any markdown codeblocks."
    )

    user_prompt = (
        f"Please analyze this column profile:\n\n{stats_text}\n"
        f"Optional Dataset Context: {dataset_context}\n"
        "Return the 'description' and 'recommendation' strictly in JSON format."
    )

    payload = {
        "model": "llama-3.1-8b-instant",
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt}
        ],
        "temperature": 0.15,
        "max_tokens": 300,
        "response_format": {"type": "json_object"}
    }

    try:
        response = _post_with_retry(url, headers=headers, json_payload=payload)
        if response.status_code == 200:
            result_json = response.json()
            content_str = result_json["choices"][0]["message"]["content"]
            parsed_content = json.loads(content_str)
            return {
                "description": parsed_content.get("description", "No description generated."),
                "recommendation": parsed_content.get("recommendation", "No recommendations found.")
            }
        else:
            return {
                "description": f"Failed to generate description. Groq API returned status code {response.status_code}.",
                "recommendation": "Please verify your Groq API Key and internet connectivity."
            }
    except Exception as e:
        return {
            "description": f"AI Engine Connection Error: {str(e)}",
            "recommendation": "Review backend logs or check network state."
        }

def generate_outlier_insights(col_name, outliers_data, dataset_context=""):
    """
    Calls the Llama-3 API to generate a physical/business explanation for why these 
    specific extreme outliers might exist based on their row context.
    """
    api_key = groq_pool.current()
    if not api_key:
        return {
            "explanation": "API Key missing. Please set GROQ_API_KEY in the app/.env file to unlock AI outlier analysis."
        }

    url = "https://api.groq.com/openai/v1/chat/completions"
    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    outliers_text = json.dumps(outliers_data, indent=2)

    system_prompt = (
        "You are a world-class Principal Data Architect and Senior Data Analyst.\n"
        "Your objective is to analyze the top extreme outliers of a column along with their surrounding row context and provide a highly professional, physical/business explanation for WHY these anomalies might exist.\n"
        "Are they data entry errors, system glitches, or legitimate extreme business events? Provide your expert hypothesis.\n"
        "You must output your response in EXACT JSON format with a single field:\n"
        "1. 'explanation': A sophisticated, concise paragraph (2-3 sentences) explaining the likely root cause of these outliers based on the provided context.\n"
        "Your output must be valid JSON only, without any markdown codeblocks."
    )

    user_prompt = (
        f"Column Name: '{col_name}'\n"
        f"Top Extreme Outliers and Context Rows:\n{outliers_text}\n"
        f"Optional Dataset Context: {dataset_context}\n"
        "Return the 'explanation' strictly in JSON format."
    )

    payload = {
        "model": "llama-3.1-8b-instant",
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt}
        ],
        "temperature": 0.2,
        "max_tokens": 250,
        "response_format": {"type": "json_object"}
    }

    try:
        response = _post_with_retry(url, headers=headers, json_payload=payload)
        if response.status_code == 200:
            result_json = response.json()
            content_str = result_json["choices"][0]["message"]["content"]
            parsed_content = json.loads(content_str)
            return {
                "explanation": parsed_content.get("explanation", "No explanation generated.")
            }
        else:
            return {
                "explanation": f"Failed to generate explanation. Groq API returned status code {response.status_code}."
            }
    except Exception as e:
        return {
            "explanation": f"AI Engine Connection Error: {str(e)}"
        }


async def generate_column_insights_async(col_name: str, profile_details: dict, dataset_context: str = "") -> dict:
    """
    Native async version of generate_column_insights using httpx.AsyncClient.
    Allows true async concurrent Groq calls without thread pool overhead.
    Automatically rotates to the next pool key on 429.
    """
    import httpx

    api_key = groq_pool.current()
    if not api_key:
        return {
            "description": "API Key missing. Please set GROQ_API_KEY in the app/.env file.",
            "recommendation": "Set your Groq API key to unlock expert AI descriptions."
        }

    # Build the same statistical profile text as the sync version
    stats_text = (
        f"Column Name: '{col_name}'\n"
        f"Basic Pandas Type: {profile_details.get('pandas_dtype')}\n"
        f"Inferred Semantic Type: {profile_details.get('semantic_type')}\n"
        f"Completeness: {profile_details.get('non_null_count')} valid records, "
        f"{profile_details.get('null_percentage')}% missing.\n"
        f"Cardinality: {profile_details.get('unique_count')} unique values ({profile_details.get('unique_ratio')}% ratio).\n"
        f"Sample values: [{profile_details.get('sample_data')}]\n"
    )

    if profile_details.get("mean") is not None:
        stats_text += (
            f"Numeric Distribution: Min={profile_details.get('min')}, Max={profile_details.get('max')}, "
            f"Mean={profile_details.get('mean')}, StdDev={profile_details.get('std')}.\n"
            f"Anomalies: {profile_details.get('outliers_count')} outliers detected via Interquartile Range (IQR).\n"
        )

    system_prompt = (
        "You are a world-class Principal Data Architect and Senior Data Analyst with 15+ years of experience in enterprise business intelligence, database indexing, and data governance.\n"
        "Your objective is to analyze a column's statistical profile, cardinality, null ratios, anomalies, and sample values, and write authoritative, elite-level business summaries.\n"
        "You must output your response in EXACT JSON format with these two fields:\n"
        "1. 'description': A highly professional, single-sentence executive business definition.\n"
        "2. 'recommendation': An expert data quality, cleaning, and optimization recommendation.\n"
        "Your output must be valid JSON only, without any markdown codeblocks."
    )

    user_prompt = (
        f"Please analyze this column profile:\n\n{stats_text}\n"
        f"Optional Dataset Context: {dataset_context}\n"
        "Return the 'description' and 'recommendation' strictly in JSON format."
    )

    payload = {
        "model": "llama-3.1-8b-instant",
        "messages": [
            {"role": "system", "content": system_prompt},
            {"role": "user", "content": user_prompt}
        ],
        "temperature": 0.15,
        "max_tokens": 300,
        "response_format": {"type": "json_object"}
    }

    headers = {
        "Authorization": f"Bearer {api_key}",
        "Content-Type": "application/json"
    }

    for attempt in range(3):
        try:
            async with httpx.AsyncClient(timeout=10.0) as client:
                response = await client.post(
                    "https://api.groq.com/openai/v1/chat/completions",
                    headers=headers,
                    json=payload
                )
                if response.status_code == 429:
                    headers["Authorization"] = f"Bearer {groq_pool.rotate()}"
                    await asyncio.sleep(2 ** attempt)
                    continue
                response.raise_for_status()
                content_str = response.json()["choices"][0]["message"]["content"]
                parsed = json.loads(content_str)
                return {
                    "description": parsed.get("description", "No description generated."),
                    "recommendation": parsed.get("recommendation", "No recommendations found.")
                }
        except Exception:
            if attempt == 2:
                return {
                    "description": "Could not generate AI description.",
                    "recommendation": "AI engine temporarily unavailable. Check logs."
                }
            await asyncio.sleep(2 ** attempt)

    return {"description": "Could not generate description.", "recommendation": ""}

File: src/lib/describer.ts
/**
 * describer.ts — Groq AI column description generator
 * Port of describer.py — same prompts, same JSON output format
 */

import type { ColumnProfile, ExecutiveReport, DatasetProfile } from "@/types";
import { getCurrentKey, rotateKey } from "./groqPool";

const GROQ_URL = "https://api.groq.com/openai/v1/chat/completions";
const MODEL = "llama-3.1-8b-instant";

async function groqPost(
  messages: { role: string; content: string }[],
  maxTokens = 350,
  jsonMode = true
): Promise<string | null> {
  let apiKey = getCurrentKey();
  if (!apiKey) return null;

  for (let attempt = 0; attempt < 3; attempt++) {
    try {
      const res = await fetch(GROQ_URL, {
        method: "POST",
        headers: {
          "Authorization": `Bearer ${apiKey}`,
          "Content-Type": "application/json",
        },
        body: JSON.stringify({
          model: MODEL,
          messages,
          temperature: 0.15,
          max_tokens: maxTokens,
          ...(jsonMode ? { response_format: { type: "json_object" } } : {}),
        }),
        signal: AbortSignal.timeout(12000),
      });

      if (res.status === 429) {
        apiKey = rotateKey();
        await new Promise((r) => setTimeout(r, 1000 * (attempt + 1)));
        continue;
      }

      if (!res.ok) return null;
      const data = await res.json();
      return data.choices?.[0]?.message?.content ?? null;
    } catch {
      if (attempt < 2) await new Promise((r) => setTimeout(r, 1000 * (attempt + 1)));
    }
  }
  return null;
}

export async function generateColumnInsights(
  colName: string,
  profile: ColumnProfile,
  datasetContext = ""
): Promise<{ description: string; recommendation: string }> {
  const statsText = [
    `Column Name: '${colName}'`,
    `Basic JS Type: ${profile.pandas_dtype}`,
    `Inferred Semantic Type: ${profile.semantic_type}`,
    `Completeness: ${profile.non_null_count} valid records, ${profile.null_percentage}% missing.`,
    `Cardinality: ${profile.unique_count} unique values (${profile.unique_ratio}% ratio).`,
    `Sample values: [${profile.sample_data}]`,
    profile.mean !== null
      ? `Numeric Distribution: Min=${profile.min}, Max=${profile.max}, Mean=${profile.mean}, StdDev=${profile.std}.`
      : "",
    profile.outliers_count > 0
      ? `Anomalies: ${profile.outliers_count} outliers detected via IQR.`
      : "",
  ]
    .filter(Boolean)
    .join("\n");

  const systemPrompt = `You are a world-class Principal Data Architect with 15+ years of experience in enterprise BI and data governance.
Analyze a column's statistical profile and write authoritative, elite-level business summaries.
Output EXACT JSON with two fields:
1. 'description': A highly professional, single-sentence executive business definition explaining why this data matters.
2. 'recommendation': An expert data quality, cleaning, and optimization recommendation with concrete technical instructions.
Output valid JSON only, no markdown codeblocks.`;

  const userPrompt = `Analyze this column profile:\n\n${statsText}\n\nOptional Dataset Context: ${datasetContext}\n\nReturn 'description' and 'recommendation' in JSON.`;

  const content = await groqPost([
    { role: "system", content: systemPrompt },
    { role: "user", content: userPrompt },
  ]);

  if (!content) {
    return {
      description: "AI description temporarily unavailable.",
      recommendation: "Check API key and connectivity.",
    };
  }

  try {
    const parsed = JSON.parse(content);
    return {
      description: parsed.description ?? "No description generated.",
      recommendation: parsed.recommendation ?? "No recommendations found.",
    };
  } catch {
    return { description: content.slice(0, 200), recommendation: "" };
  }
}

export async function generateOutlierInsights(
  colName: string,
  outliers: ColumnProfile["top_outliers"],
  datasetContext = ""
): Promise<{ explanation: string }> {
  const systemPrompt = `You are a world-class Principal Data Architect.
Analyze the top extreme outliers of a column and provide a highly professional physical/business explanation for WHY these anomalies might exist.
Output EXACT JSON with one field: 'explanation': A sophisticated 2-3 sentence paragraph.
Output valid JSON only, no markdown codeblocks.`;

  const userPrompt = `Column Name: '${colName}'\nTop Extreme Outliers:\n${JSON.stringify(outliers, null, 2)}\nDataset Context: ${datasetContext}\n\nReturn 'explanation' in JSON.`;

  const content = await groqPost(
    [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ],
    280
  );

  if (!content) return { explanation: "Could not generate an explanation." };

  try {
    const parsed = JSON.parse(content);
    return { explanation: parsed.explanation ?? "No explanation generated." };
  } catch {
    return { explanation: content.slice(0, 300) };
  }
}

export async function generateExecutiveReport(
  profile: DatasetProfile,
  datasetContext = ""
): Promise<ExecutiveReport> {
  const schemaSummary = profile.columns
    .slice(0, 30)
    .map(
      (c) =>
        `- ${c.name} (${c.semantic_type}, ${c.null_percentage}% null, ${c.unique_count} unique)`
    )
    .join("\n");

  const systemPrompt = `You are a world-class Chief Data Officer writing an executive data governance report.
Analyze the provided dataset schema statistics and generate a comprehensive business intelligence brief.
Output EXACT JSON with these fields:
- domain_label: 1-3 word domain classification (e.g. "E-Commerce Sales", "Healthcare Records")
- business_overview: 2-3 sentence executive summary of what this dataset represents
- key_findings: Array of 3-5 specific data quality findings (strings)
- governance_scope: 1-2 sentences on data governance scope and responsibilities
- health_assessment: 1-2 sentences on overall data health and readiness
- recommendations: Array of 3-5 specific actionable recommendations (strings)
Output valid JSON only, no markdown.`;

  const userPrompt = `Dataset: ${profile.filename ?? "Unknown"}
Total Rows: ${profile.total_rows.toLocaleString()}
Total Columns: ${profile.total_cols}
Duplicate Rows: ${profile.duplicate_rows}
Data Completeness: ${profile.completeness}%
Health Score: ${profile.health_score}/100 (${profile.quality_label})
Anomaly Penalty: ${profile.anomaly_penalty}

Column Schema:
${schemaSummary}

Context: ${datasetContext}

Generate a comprehensive executive intelligence report.`;

  const content = await groqPost(
    [
      { role: "system", content: systemPrompt },
      { role: "user", content: userPrompt },
    ],
    600,
    true
  );

  if (!content) {
    return {
      domain_label: "Unknown Domain",
      business_overview: "Executive report generation failed. Please try again.",
      key_findings: ["Analysis unavailable"],
      governance_scope: "Not determined",
      health_assessment: `Health score: ${profile.health_score}/100`,
      recommendations: ["Check API connectivity"],
    };
  }

  try {
    return JSON.parse(content) as ExecutiveReport;
  } catch {
    return {
      domain_label: "Data Asset",
      business_overview: content.slice(0, 400),
      key_findings: [],
      governance_scope: "",
      health_assessment: `Health score: ${profile.health_score}/100`,
      recommendations: [],
    };
  }
}

export async function generateChatResponse(
  schemaText: string,
  userMessage: string
): Promise<{ code: string | null; answer: string | null }> {
  const codePrompt = `You are an expert JavaScript data analyst. The user's dataset is in an array of objects called \`data\`.
Schema:
${schemaText}

Write JavaScript code to answer: "${userMessage}"
Rules:
1. Store final answer in a variable called \`result\`. The \`result\` MUST be a friendly, natural language string directly answering the user (e.g. \`let result = 'The average is ' + avg + '.';\`). Do NOT just return a JSON object or array.
2. Use only native JS array methods (map, filter, reduce, etc.)
3. Output ONLY raw JS code — no imports, no markdown, no console.log
4. Do NOT use eval, require, or any Node.js APIs
5. Column names might contain spaces, uppercase letters, or special characters. Always access properties on row objects using bracket notation with the exact column name as specified in the schema (e.g. \`row['Court Name']\` or \`row['Case No']\`), rather than dot notation. Do not assume normalized or camelCase keys.`;

  const codeContent = await groqPost(
    [{ role: "user", content: codePrompt }],
    500,
    false
  );

  return { code: codeContent, answer: null };
}
